sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/SqlAnalyzer.java

```diff
@@ -107,6 +128,43 @@ ResolvedStatement analyze(String sql) {
     return Analyzer.analyzeStatement(sql, options, catalog);
   }
 
+  /**
+   * Accepts the ParseResumeLocation for the current position in the SQL string. Advances the
+   * ParseResumeLocation to the start of the next statement. Adds user-defined functions to the
+   * catalog for use in following statements. Returns the resolved AST.
+   */
+  ResolvedStatement analyzeNextStatement(
+      ParseResumeLocation parseResumeLocation, AnalyzerOptions options, SimpleCatalog catalog) {
+    ResolvedStatement resolvedStatement =
+        Analyzer.analyzeNextStatement(parseResumeLocation, options, catalog);
+    if (resolvedStatement.nodeKind() == RESOLVED_CREATE_FUNCTION_STMT) {
+      ResolvedCreateFunctionStmt createFunctionStmt =
+          (ResolvedCreateFunctionStmt) resolvedStatement;
+      Function userFunction =
+          new Function(
+              createFunctionStmt.getNamePath(),
+              USER_DEFINED_FUNCTIONS,
+              // TODO(BEAM-9954) handle aggregate functions
+              // TODO(BEAM-9969) handle table functions
+              Mode.SCALAR,
+              com.google.common.collect.ImmutableList.of(createFunctionStmt.getSignature()));
+      try {
+        catalog.addFunction(userFunction);
```

Comment:
I wasn't aware that this addFunction can throw an exception. Any example that could fail this line? E.g. adding a function with duplicated name?

Reply:
Yes. See https://github.com/apache/beam/pull/12153/files#diff-bf7c0f0293cfabc387959dbd4494a5cfR2855-R2862